packages/nextjs/hooks/scaffold-eth/useDeployedContractInfo.ts

```diff
@@ -1,52 +1,52 @@
-import { useEffect, useState } from "react";
+import { useEffect, useMemo, useState } from "react";
 import { useProvider } from "wagmi";
 import { getTargetNetwork } from "~~/utils/scaffold-eth";
-type GeneratedContractType = {
-  address: string;
-  abi: any[];
-};
+import contracts from "../../generated/hardhat_contracts";
+import { Contract, ContractName, ContractCodeStatus, DefaultChain } from "./contract.types";
 
 /**
  * @dev use this hook to get a deployed contract from `yarn deploy` generated files.
  * @param contractName - name of deployed contract
  * @returns {GeneratedContractType | undefined} object containing contract address and abi or undefined if contract is not found
  */
-export const useDeployedContractInfo = (contractName: string | undefined | null) => {
+export const useDeployedContractInfo = <TContractName extends ContractName>(contractName: TContractName) => {
   const configuredChain = getTargetNetwork();
-  const [deployedContractData, setDeployedContractData] = useState<undefined | GeneratedContractType>(undefined);
-  const [isLoading, setIsLoading] = useState(true);
+  const deployedContract = useMemo(
+    () =>
+      contracts[`${configuredChain.id}` as DefaultChain]?.[0]?.contracts?.[
+        contractName as ContractName
+      ] as Contract<TContractName>,
+    [configuredChain.id, contractName],
+  );
+  const [status, setStatus] = useState<ContractCodeStatus>(ContractCodeStatus.LOADING);
   const provider = useProvider({ chainId: configuredChain.id });
 
   useEffect(() => {
-    const getDeployedContractInfo = async () => {
-      setIsLoading(true);
-      let ContractData;
-      try {
-        ContractData = require("~~/generated/hardhat_contracts.json");
-        const contractsAtChain = ContractData[configuredChain.id as keyof typeof ContractData];
-        const contractsData = contractsAtChain?.[0]?.contracts;
-        const deployedContract = contractsData?.[contractName as keyof typeof contractsData];
+    let mounted = true;
 
-        if (!deployedContract || !contractName || !provider) {
-          return;
-        }
+    const checkContractDeployment = async () => {
+      const code = await provider.getCode((deployedContract as Contract<TContractName>).address);
 
-        const code = await provider.getCode(deployedContract.address);
-        // If contract code is `0x` => no contract deployed on that address
-        if (code === "0x" || !contractsData || !(contractName in contractsData)) {
-          return;
-        }
-        setDeployedContractData(contractsData[contractName]);
-      } catch (e) {
-        // Contract not deployed or file doesn't exist.
-        setDeployedContractData(undefined);
-      } finally {
-        setIsLoading(false);
+      if (!mounted) {
+        return;
+      }
+      // If contract code is `0x` => no contract deployed on that address
+      if (code === "0x" || !deployedContract) {
```

Comment:
I think `!deployedContract` condition should be put before `line 28` because it error's out when if people forgot to do `yarn deploy` I think we discussed on other PR that app should not break if user forgots it not able to find it right now 😅. 

Also I think maybe can update this file in a different PR 🤔? Since it has been used at different places we need to test it properly which increases the overhead of testing other stuff of this PR, 

btw I love the changes but need to test it properly, Thanks 🙌

Reply:
You're right about the check needing to happen before :+1: 

As for other places that use this hook, they've all been checked and updated in this PR (since the other hooks need this as well). Not sure if splitting it in a different PR is needed, since these changes require testing all the same things anyway, I believe.

Reply:
Yup makes sense 🙌